src/execution/mod.rs

```diff
@@ -263,15 +264,41 @@ where
                                 let address = address as usize;
                                 mem.data.get(address..(address + 4))
                             })
-                            .expect("TODO trap here");
-
-                        let data: [u8; 4] = data.try_into().expect("this to be exactly 4 bytes");
+                            .and_then(|slice| slice.try_into().ok())
+                            .expect("this to be exactly 4 bytes");
                         u32::from_le_bytes(data)
                     };
 
                     stack.push_value(Value::I32(data));
                     trace!("Instruction: i32.load [{relative_address}] -> [{data}]");
                 }
+                F32_LOAD => {
+                    let memarg = MemArg::read_unvalidated(&mut wasm);
+                    let relative_address: u32 =
+                        stack.pop_value(ValType::NumType(NumType::I32)).into();
+
+                    let mem = self.store.mems.first().unwrap_validated(); // there is only one memory allowed as of now
+
+                    let data: f32 = {
+                        // The spec states that this should be a 33 bit integer
+                        // See: https://webassembly.github.io/spec/core/syntax/instructions.html#memory-instructions
+                        let _address = memarg.offset.checked_add(relative_address);
+                        let data = memarg
+                            .offset
+                            .checked_add(relative_address)
+                            .and_then(|address| {
+                                let address = address as usize;
+                                mem.data.get(address..(address + 4))
+                            })
+                            .and_then(|slice| slice.try_into().ok())
+                            .expect("this to be exactly 4 bytes");
```

Comment:
This is not correct because it panics when an invalid memory location has been accessed. In that case it should instead trap (or panic with "TODO trap here" like previously).

I suggest the following:
```suggestion
                        let data = memarg
                            .offset
                            .checked_add(relative_address)
                            .and_then(|address| {
                                let address = address as usize;
                                mem.data.get(address..(address + 4))
                                    .map(|slice| slice.try_into().expect("this to be exactly 4 bytes"))
                            })
                            .expect("TODO trap here");
```

---
Even better would be this abomination (see [first_chunk](https://doc.rust-lang.org/std/primitive.slice.html#method.first_chunk)), which gets rid of the expect entirely:
```suggestion
                        let data = memarg
                            .offset
                            .checked_add(relative_address)
                            .and_then(|address| {
                                let address = address as usize;
                                mem.data.get(address..)
                            })
                            .and_then(<[u8]>::first_chunk::<4>)
                            .expect("TODO trap here");
```

Reply:
My bad! Did it the first way, though, hope it's good now.

Reply:
👍 

I will probably create another PR to do the same for all other places where we have to read a `[u8; 4]` from memory.